### DISCUSSION TITLE
Additional register information for Decimals

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#365] [Q&A]
I separate your question/request in a new thread: 
```
Also I think we should have in the register definition, a significant digits:

 decNumber *decimal_value;
 int sig_digits;
```
This is what I use now:
```
register->decimal_value = address of allocated decimal memory     
register->string_length = length of allocated memory  // acceptable as decimal value is a kind of string
``` 
I thought about keeping the significant digits, but I couldn't find an attribute where I could push it. So either please add a new one or overlay an attribute that we don't need for decimals.

**Comment by adesutherland** [comment]
Remember that a register can have a string value and a decimal number quite independently - therefore, you should not use the same length (or buffer) for both of them :-)

**Comment by adesutherland** [comment]
In the decnumber.pdf page 31 we have

> For many applications, there will be a known maximum length for numbers and
> DECNUMDIGITS can be set to that length, as in Example 1 (see page 11). In others, the
> length may vary over a wide range and it then becomes the programmer’s responsibility
> to ensure that there are sufficient units available immediately following the decNumber
> lsu field. This can be achieved by enclosing the decNumber in other structures which
> append various lengths of unit arrays, or in the more general case by allocating storage
> with sufficient space for the other decNumber fields and the units of the number.

So I think in the register we need something like this

    DecNumber *decimal_value;
    size_r decimal_buffer_length;

and allocation logic is like

    decimal_buffer_length = sizeof(DecNumber) + calculated_buffer_size_depending_on_numdigits;
    decimal_value = malloc(decimal_buffer_length);

Later logic may decide if the buffer needs to be reallocated to make it bigger;

**Reply by Peter-Jacob** [reply]
This is true! When you add decimal_buffer_length, don't forget to add something like decimal_digit_numbers.
